pkgs/development/compilers/gcc/common/pre-configure.nix

```diff
@@ -77,7 +77,14 @@ in lib.optionalString (hostPlatform.isSunOS && hostPlatform.is64bit) ''
 #  - If GCC allowed implicit declaration of symbols, it would not fail during
 #    compilation even if the configure scripts did not check header presence.
 #
-+ lib.optionalString (hostPlatform.isDarwin) ''
+# NOTE 2023/4/15: Unfortunately, making this fix depend on hostPlatform
+# makes binaries cross-compiled from Darwin to Linux different from binaries
+# built on Linux, cross-compiled or not. Additionally, newer macOS versions
+# include aligned_alloc in the standard library headers.
+# Depend on targetPlatform instead, keeping non-cross gcc compatible with macOS
+# 10.12. Cross-compiling gcc output the same binaries as Linux gcc but
+# require macOS 10.15.
++ lib.optionalString (targetPlatform.isDarwin) ''
```

Comment:
I think it's a bit more tricky: `aligned_alloc` is `gcc`'s builtin. As a result I expect it to be used both in `gcc`'s binary and target libraries built by `gcc`. That means you likely want `ac_cv_func_aligned_alloc=no` for `--host=darwin` (and any `--target=`) to keep the properties of `gcc` binary. And for `--target=darwin` to keep the properties of target libraries like `libgomp`.

In theory it would be nice if `gcc`'s build system provided hooks to affect only host or only target with `ac_cv_func_aligned_alloc=no`. But it's a bit tricky to get right. I think the canonical way to do it is to pass it via `STAGE{2,3}_CONFIGURE_FLAGS` to `gcc`.

Getting back to your issue you might get a reasonable outcome with `lib.optionalString (hostPlatform.isDarwin || targetPlatform.isDarwin)`.

Reply:
> Getting back to your issue you might get a reasonable outcome with lib.optionalString (hostPlatform.isDarwin || targetPlatform.isDarwin).

So

```
lib.optionalString (hostPlatform.isDarwin || targetPlatform.isDarwin) ''
      export ac_cv_func_aligned_alloc=no
''
```

?

Isn't your new condition a super-set of the old, because of the `||`? My issue is that without this PR (or similar) cross-compiled binaries for Raspberry Pi don't reference `aligned_alloc` whereas cross-compiled binaries built on Linux do. It seems to me your suggestion strictly expands the number of configurations where `aligned_alloc` will be left out.

Did you perhaps mean `lib.optionalString (hostPlatform.isDarwin && targetPlatform.isDarwin)`, with `&&` instead of `||`?

Reply:
My suggestion was based on the following goals:
- when building on darwin build (either for linux or for darwin target) there should be no references to `aligned_allos)
- when building for target darwin (either on linux or darwin build) there should be no references to `aligned_allos)

Hence the `||`.

Reply:
But I want to have darwin building for linux result in the same binaries as when linux builds for linux. That is, I want the `aligned_alloc` reference in the binaries because that's what linux building for linux have.

Do you know of a way to achieve that, without breaking darwin?

Reply:
Reading through `gcc/doc/install.texi` I found this bit (TIL!):

```
@subsubheading Overriding @command{configure} test results

Sometimes, it might be necessary to override the result of some
@command{configure} test, for example in order to ease porting to a new
system or work around a bug in a test.  The toplevel @command{configure}
script provides three variables for this:

@table @code

@cindex @code{build_configargs}
@item build_configargs
The contents of this variable is passed to all build @command{configure}
scripts.

@cindex @code{host_configargs}
@item host_configargs
The contents of this variable is passed to all host @command{configure}
scripts.

@cindex @code{target_configargs}
@item target_configargs
The contents of this variable is passed to all target @command{configure}
scripts.

@end table
```

Thus most robust way to get that behavior might be the (untested):

```nix
lib.optionalString (buildPlatform.isDarwin) ''
      export build_configargs=ac_cv_func_aligned_alloc=no
'' ++ lib.optionalString (hostPlatform.isDarwin) ''
      export host_configargs=ac_cv_func_aligned_alloc=no
'' ++ lib.optionalString (targetPlatform.isDarwin) ''
      export target_configargs=ac_cv_func_aligned_alloc=no
''
```

Reply:
Thank you very much! Separate configure flags seem to be the correct fix, and achieves my goal og build paltform independent binaries. PTAL.